pymc_marketing/mmm/fourier.py

```diff
@@ -0,0 +1,653 @@
+        Number of days in a period.
+    prefix : str
+        Name of model coordinates
+    default_prior : Prior
+        Default prior distribution for the fourier seasonality
+        beta parameters.
+    """
+
+    days_in_period: float
+    prefix: str = "fourier"
+
+    default_prior = Prior("Laplace", mu=0, b=1)
+
+    def __init__(
+        self,
+        n_order: int,
+        prefix: str | None = None,
+        prior: Prior | None = None,
+    ) -> None:
+        self.n_order = n_order
+        self.prefix = prefix or self.prefix
+        self.prior = prior or self.default_prior
+
+        if not self.prior.dims:
+            self.prior = self.prior.deepcopy()
+            self.prior.dims = self.prefix
+
+        if self.prefix not in self.prior.dims:
+            raise ValueError(f"Prior distribution must have dimension {self.prefix}")
+
+    @property
+    def nodes(self) -> list[str]:
+        """Fourier node names for model coordinates."""
+        return [
+            f"{func}_{i}" for func in ["sin", "cos"] for i in range(1, self.n_order + 1)
+        ]
+
+    @property
+    def variable_name(self) -> str:
+        """Name of variable that multiplies the fourier modes."""
+        return f"{self.prefix}_beta"
+
+    def apply(
+        self,
+        dayofyear: pt.TensorLike,
+        result_callback: Callable[[pt.TensorVariable], None] | None = None,
+    ) -> pt.TensorVariable:
+        """Apply fourier seasonality to day of year.
+
+        Must be used within a PyMC model context.
+
+        Parameters
+        ----------
+        dayofyear : pt.TensorLike
+            Day of year.
+        result_callback : Callable[[pt.TensorVariable], None], optional
+            Callback function to apply to the result, by default None
+
+        Returns
+        -------
+        pt.TensorVariable
+            Fourier seasonality
+
+        Examples
+        --------
+        Save off the result before summing through the prefix dimension.
+
+        .. code-block:: python
+
+            import pandas as pd
+
+            import pymc as pm
+
+            from pymc_marketing.mmm import YearlyFourier
+
+            fourier = YearlyFourier(n_order=3)
+
+            def callback(result):
+                pm.Deterministic("fourier_trend", result, dims=("date", "fourier"))
+
+            dates = pd.date_range("2023-01-01", periods=52, freq="W-MON")
+
+            coords = {
+                "date": dates,
+            }
+            with pm.Model(coords=coords) as model:
+                dayofyear = dates.dayofyear.to_numpy()
+                fourier.apply(dayofyear, result_callback=callback)
+
+        """
+        periods = dayofyear / self.days_in_period
+
+        model = pm.modelcontext(None)
+        model.add_coord(self.prefix, self.nodes)
+
+        beta = self.prior.create_variable(self.variable_name)
+
+        fourier_modes = generate_fourier_modes(periods=periods, n_order=self.n_order)
+
+        DUMMY_DIM = "DATE"
```

Comment:
Is there any reason for this string to be the upper case?

Reply:
No reason. It can be anything. A name is just required for the `dim_handler`. This is just associated the dimension of the `dayofyear` input vector